scripts/base/protocols/ldap/main.zeek

```diff
@@ -0,0 +1,513 @@
+                                                       LDAP::ProtocolOpcode_EXTENDED_RESPONSE };
+
+global OPCODES_SEARCH: set[LDAP::ProtocolOpcode] = { LDAP::ProtocolOpcode_SEARCH_REQUEST,
+                                                     LDAP::ProtocolOpcode_SEARCH_RESULT_ENTRY,
+                                                     LDAP::ProtocolOpcode_SEARCH_RESULT_DONE,
+                                                     LDAP::ProtocolOpcode_SEARCH_RESULT_REFERENCE };
+
+#############################################################################
+redef record connection += {
+  ldap_proto: string &optional;
+  ldap_messages: table[int] of Message &optional;
+  ldap_searches: table[int] of Search &optional;
+};
+
+#############################################################################
+event zeek_init() &priority=5 {
+  Analyzer::register_for_ports(Analyzer::ANALYZER_LDAP_TCP, LDAP::ports_tcp);
+  Analyzer::register_for_ports(Analyzer::ANALYZER_LDAP_UDP, LDAP::ports_udp);
+
+  Log::create_stream(LDAP::LDAP_LOG, [$columns=Message, $ev=log_ldap, $path="ldap", $policy=log_policy]);
+  Log::create_stream(LDAP::LDAP_SEARCH_LOG, [$columns=Search, $ev=log_ldap_search, $path="ldap_search", $policy=log_policy_search]);
+}
+
+#############################################################################
+function set_session(c: connection, message_id: int, opcode: LDAP::ProtocolOpcode) {
+
+  if (! c?$ldap_messages )
+    c$ldap_messages = table();
+
+  if (! c?$ldap_searches )
+    c$ldap_searches = table();
+
+  if ((opcode in OPCODES_SEARCH) && (message_id !in c$ldap_searches)) {
+    c$ldap_searches[message_id] = [$ts=network_time(),
+                                   $uid=c$uid,
+                                   $id=c$id,
+                                   $message_id=message_id,
+                                   $result_count=0];
+
+  } else if ((opcode !in OPCODES_SEARCH) && (message_id !in c$ldap_messages)) {
+    c$ldap_messages[message_id] = [$ts=network_time(),
+                                   $uid=c$uid,
+                                   $id=c$id,
+                                   $message_id=message_id];
+  }
+
+}
+
+#############################################################################
+@if (Version::at_least("5.2.0"))
+event analyzer_confirmation_info(atype: AllAnalyzers::Tag, info: AnalyzerConfirmationInfo) {
+  if ( atype == Analyzer::ANALYZER_LDAP_TCP ) {
+    info$c$ldap_proto = "tcp";
+  }
+}
+@else @if (Version::at_least("4.2.0"))
+event analyzer_confirmation(c: connection, atype: AllAnalyzers::Tag, aid: count) {
+@else
+event protocol_confirmation(c: connection, atype: Analyzer::Tag, aid: count) {
+@endif
+
+  if ( atype == Analyzer::ANALYZER_LDAP_TCP ) {
+    c$ldap_proto = "tcp";
+  }
+
+}
+@endif
+#############################################################################
+event LDAP::message(c: connection,
+                    message_id: int,
```

Comment:
Quick googling indicates message_id is not negative, so this could be a count. Do you know why it wasn't/isn't one?

Reply:
Unsure, but it looks like we only need half the range of `uint32` (though our `int` range is big enough). I tried changing it, but since this value comes from ASN.1 this is more than simply swapping the types. I'd prefer to leave this as a follow-up if it ever becomes relevant.